sdk/python/tests/integration/offline_store/test_universal_historical_retrieval.py

```diff
@@ -0,0 +1,309 @@
+        end_time = datetime.utcnow()
+        print(
+            str(f"\nTime to execute job_from_sql.to_df() = '{(end_time - start_time)}'")
+        )
+
+        assert sorted(expected_df.columns) == sorted(
+            actual_df_from_sql_entities.columns
+        )
+
+        actual_df_from_sql_entities = (
+            actual_df_from_sql_entities[expected_df.columns]
+            .sort_values(by=[event_timestamp, "order_id", "driver_id", "customer_id"])
+            .drop_duplicates()
+            .reset_index(drop=True)
+        )
+        expected_df = (
+            expected_df.sort_values(
+                by=[event_timestamp, "order_id", "driver_id", "customer_id"]
+            )
+            .drop_duplicates()
+            .reset_index(drop=True)
+        )
+
+        assert_frame_equal(
+            actual_df_from_sql_entities, expected_df, check_dtype=False,
+        )
+
+        table_from_sql_entities = job_from_sql.to_arrow()
+        df_from_sql_entities = (
+            table_from_sql_entities.to_pandas()[expected_df.columns]
+            .sort_values(by=[event_timestamp, "order_id", "driver_id", "customer_id"])
+            .drop_duplicates()
+            .reset_index(drop=True)
+        )
+        assert_frame_equal(actual_df_from_sql_entities, df_from_sql_entities)
+
+    # timestamp_column = (
+    #     "e_ts"
+    #     if environment.test_repo_config.infer_event_timestamp_col
+    #     else DEFAULT_ENTITY_DF_EVENT_TIMESTAMP_COL
+    # )
+
+    # entity_df_query_with_invalid_join_key = (
+    #     f"select order_id, driver_id, customer_id as customer, "
+    #     f"order_is_success, {timestamp_column}, FROM {environment.orders_table()}"
+    # )
+    # # Rename the join key; this should now raise an error.
+    # assertpy.assert_that(store.get_historical_features).raises(
+    #     errors.FeastEntityDFMissingColumnsError
+    # ).when_called_with(
+    #     entity_df=entity_df_query_with_invalid_join_key,
+    #     features=[
+    #         "driver_stats:conv_rate",
+    #         "driver_stats:avg_daily_trips",
+    #         "customer_profile:current_balance",
+    #         "customer_profile:avg_passenger_count",
+    #         "customer_profile:lifetime_trip_count",
+    #     ],
+    # )
+
+    # Rename the join key; this should now raise an error.
+    # orders_df_with_invalid_join_key = orders_df.rename(
+    #     {"customer_id": "customer"}, axis="columns"
+    # )
+    # assertpy.assert_that(store.get_historical_features).raises(
+    #     errors.FeastEntityDFMissingColumnsError
+    # ).when_called_with(
+    #     entity_df=orders_df_with_invalid_join_key,
+    #     features=[
+    #         "driver_stats:conv_rate",
+    #         "driver_stats:avg_daily_trips",
+    #         "customer_profile:current_balance",
+    #         "customer_profile:avg_passenger_count",
+    #         "customer_profile:lifetime_trip_count",
+    #     ],
+    # )
+
+    job_from_df = store.get_historical_features(
```

Comment:
thoughts on splitting sql vs df into two separate tests with the same initial setup? would make it so one can distinguish from test failures whether their bug is in common logic or logic specific to sql vs df

Reply:
I think a single test is better - sql and df params are testing the same API roughly, and tests will run faster if there's fewer separate tests.